### DISCUSSION TITLE
Handle backtracking via transactional memory-ish constructs?

### DISCUSSION THREAD

**Discussion by stevefan1999-personal** [kevinmehall/rust-peg#301] [Ideas]
I was thinking about adding a set of API for "commit" and "rollback" for the symbol table: it will have a set of "snapshots" and when we kick back our parser to a certain point,. "rollback" action occurs reverting the symbol table to a previously state. 

That way, we can properly backtrack without having ghost variables in the temporary symbol table. I'm not certain what this is called but I think there are people implemented things like this before, but the closest thing I can think of is transactional memory.

**Comment by kevinmehall** [comment]
This sounds like the [state feature in Pegasus (C#)](https://github.com/otac0n/Pegasus/wiki/Significant-Whitespace-Parsing). There's some discussion of that on https://github.com/kevinmehall/rust-peg/issues/223.
